File: python/hexabloom.py
import bitarray

# intdigest returns the result as an integer, which is easier to work with compared to a btyes object
from xxhash import xxh3_64_intdigest

class BloomFilterClient:
    def __init__(self, bitmap, number_of_hashes: int):
        self.bitmap = bitmap
        self.k = number_of_hashes
        self.M = len(bitmap)

    def contains_str(self, value: str) -> bool:
        return self.contains_bytes(value.encode("utf-8"))

    def contains_bytes(self, value: bytes) -> bool:
        # pay attention to the seed values and the hash function usef
        hash1 = xxh3_64_intdigest(value, seed=0) % self.M
        hash2 = xxh3_64_intdigest(value, seed=32) % self.M
        res = self.bitmap[hash1]
        for i in range(self.k):
            mo = (hash1 + i * hash2) % self.M
            if not self.bitmap[mo]:
                return False
        return res

    @classmethod
    def from_file_with_hashes(cls, file_path: str) -> "BloomFilterClient":
        with open(file_path, "rb") as f:
            hashes = int.from_bytes(f.read(4), "big")
            bitmap = bitarray.bitarray(endian="little")
            bitmap.fromfile(f)
            return cls(bitmap, hashes)

File: go/hexabloom.go
package hexabloom

import (
	"math/big"
	"os"

	"github.com/zeebo/xxh3"
)

type BloomFilterClient struct {
	bitmap *big.Int
	k      int
	M      int
}

func (bfc *BloomFilterClient) ContainsStr(value string) bool {
	return bfc.ContainsBytes([]byte(value))
}

func (bfc *BloomFilterClient) ContainsBytes(value []byte) bool {
	// pay attention to the seed values and the hash function used
	hash1 := int(xxh3.HashSeed(value, 0) % uint64(bfc.M))
	hash2 := int(xxh3.HashSeed(value, 32) % uint64(bfc.M))
	for i := 0; i < bfc.k; i++ {
		mo := int((hash1 + i*hash2) % bfc.M)
		bit := bfc.bitmap.Bit(mo)
		if bit == 0 {
			return false
		}
	}
	return true
}

func NewBloomFilterClientFromFile(filePath string) (*BloomFilterClient, error) {
	fileBytes, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}
	numberOfHashesBytes := fileBytes[:4]
	bitmap := fileBytes[4:]

	// convert byte endianness
	numberOfHashes := int(numberOfHashesBytes[3]) | int(numberOfHashesBytes[2])<<8 | int(numberOfHashesBytes[1])<<16 | int(numberOfHashesBytes[0])<<24
	for i, j := 0, len(bitmap)-1; i < j; i, j = i+1, j-1 {
		bitmap[i], bitmap[j] = bitmap[j], bitmap[i]
	}

	// create bloom filter client
	return &BloomFilterClient{
		bitmap: new(big.Int).SetBytes(bitmap),
		k:      numberOfHashes,
		M:      len(bitmap) * 8,
	}, nil

}
